runtime/src/main/scala/akka/grpc/internal/Marshallers.scala

```diff
@@ -18,19 +18,35 @@ abstract class BaseMarshaller[T](val protobufSerializer: ProtobufSerializer[T])
     extends io.grpc.MethodDescriptor.Marshaller[T]
     with WithProtobufSerializer[T] {
   override def parse(stream: InputStream): T = {
-    val baos = new ByteArrayOutputStream(math.max(64, stream.available()))
-    val buffer = new Array[Byte](32 * 1024)
+    val buffer =
+      new Array[Byte](stream match {
+        case k: KnownLength => math.max(0, k.available()) // No need to oversize this if we already know the size
+        case _              => 32 * 1024
+      })
 
     // Blocking calls underneath...
     // we can't avoid it for the moment because we are relying on the Netty's Channel API
-    var bytesRead = stream.read(buffer)
-    while (bytesRead >= 0) {
-      baos.write(buffer, 0, bytesRead)
-      bytesRead = stream.read(buffer)
+    val bytes = stream.read(buffer, 0, buffer.length) match {
+      case -1 =>
+        akka.util.ByteString.empty // EOF immediately
+      case n if n < buffer.length =>
+        // WARNING: buffer is retained in full below,
+        // which could be problematic if ProtobufSerializer.deserialize keeps a reference to the ByteString
+        akka.util.ByteString.fromArrayUnsafe(buffer, 0, n)
```

Comment:
Looking at this closer - is this reliable when the stream is not of known size? Seems like it could indeed truncate streams in that case.

Reply:
@raboof Technically, should an InputStream be allowed to return a read-length which is less than what is requested, given that it is blocking, presuming that the actual length before EOF is not shorter than the requested length?

Reply:
@raboof That said, either it should be `n <= buffer.length` or `n == buffer.length` depending on the answer to the question above.

Reply:
@raboof 

How about something like this instead:

```scala
 override def parse(stream: InputStream): T = {
    val buffer =
      new Array[Byte](stream match {
        case k: KnownLength => math.max(0, k.available()) // No need to oversize this if we already know the size
        case _              => 32 * 1024
      })

    // Blocking calls underneath...
    // we can't avoid it for the moment because we are relying on the Netty's Channel API
    val initialBytes = stream.read(buffer, 0, buffer.length)
    val nextByte = if (initialBytes < 0) -1 else stream.read() // Test for EOF
    val bytes =
      if (nextByte == -1) {
        if (initialBytes < 1) akka.util.ByteString.empty // EOF immediately
        else {
          // WARNING: buffer is retained in full below,
          // which could be problematic if ProtobufSerializer.deserialize keeps a reference to the ByteString
          akka.util.ByteString.fromArrayUnsafe(buffer, 0, initialBytes)
        }
      } else {
        val baos = new ByteArrayOutputStream(buffer.length * 2) // To avoid immediate resize
        baos.write(buffer, 0, initialBytes)
        baos.write(nextByte)

        var bytesRead = stream.read(buffer)
        while (bytesRead >= 0) {
          baos.write(buffer, 0, bytesRead)
          bytesRead = stream.read(buffer)
        }

        akka.util.ByteString.fromArrayUnsafe(baos.toByteArray)
      }

    protobufSerializer.deserialize(bytes)
  }
```

Reply:
@raboof The idea here is that we tests whether a plain single-byte read after the first chunk-read indeed is an EOF, and in that case we're already done.

Reply:
Unfortunately I think there's an issue with the akka-grpc build:

```bash
[error] (akka-grpc-runtime / update) sbt.librarymanagement.ResolveException: Error downloading com.github.ghik:silencer-lib_2.12.14:1.7.1
[error]   Not found
[error]   Not found
[error]   not found: /Users/viktorklang/.ivy2/localcom.github.ghik/silencer-lib_2.12.14/1.7.1/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/com/github/ghik/silencer-lib_2.12.14/1.7.1/silencer-lib_2.12.14-1.7.1.pom
[error]   not found: https://oss.sonatype.org/content/repositories/staging/com/github/ghik/silencer-lib_2.12.14/1.7.1/silencer-lib_2.12.14-1.7.1.pom
```

Reply:
Assuming InputStream is allowed to return a read-length which is less than what is requested (which indeed is not _entirely_ clear from the javadoc AFAICS) that seems like a safer approach, yes.

Reply:
> Unfortunately I think there's an issue with the akka-grpc build

What commit are you on? master should be using Scala 2.12.11, not 2.12.14 (#1353)

Reply:
@raboof I'm not sure what was going on, but I had to create #1373 to be able to run my build. Look at #1374 for my proposed solution.

Reply:
were you running with `sbt ++2.12.14` perhaps?

Reply:
@raboof Nope. go into master and do `project benchmarks` and then `update` and you should be getting the error as well.

Reply:
ah, you're right - I didn't spot you were going into that project. looks like that's indeed somewhat broken. Aside from updating to 2.12.14 there seem to be a couple of things to fix, #1375.